File: dragonfly/syntax/parser.py
import dragonfly
import typing

from dragonfly.output import Console

class Parser:
	def __init__(self, game: "dragonfly.Game") -> None:
		self.__game = game
		self.__showParsingProcess = False

		self.__dObjStr = ""
		self.__iObjStr = ""
		self.__parameters = ""
		self.__dObj = None
		self.__iObj = None
		self.__keyword = ""

	@property
	def game(self) -> "dragonfly.Game":
		return self.__game

	@property
	def showParsingProcess(self) -> bool:
		return self.__showParsingProcess

	@showParsingProcess.setter
	def showParsingProcess(self, val: bool) -> None:
		self.__showParsingProcess = val

	@property
	def dictionary(self) -> "dragonfly.Dictionary":
		return self.__game.dictionary

	@property
	def directObjectString(self) -> str:
		return self.__dObjStr

	@property
	def indirectObjectString(self) -> str:
		return self.__iObjStr

	@property
	def parameters(self) -> str:
		return self.__parameters

	@property
	def directObject(self) -> "dragonfly.Noun":
		return self.__dObj

	@directObject.setter
	def directObject(self, noun: "dragonfly.Noun") -> None:
		self.__dObj = noun

	@property
	def indirectObject(self) -> "dragonfly.Noun":
		return self.__iObj

	@indirectObject.setter
	def indirectObject(self, noun: "dragonfly.Noun") -> None:
		self.__iObj = noun

	@property
	def keyword(self) -> str:
		return self.__keyword

	def parse(self, line: str) -> None:

		# Gets tokens
		separated = line.strip().split(" ")
		tokens = []
		for t in separated:
			t = t.strip()
			if t: tokens.append(t)
		
		# Empty line
		if not tokens:
			self.debug("token list is empty.")
			return None

		strVerb = tokens[0]

		if strVerb.endswith(":"):
			strVerb = tokens[0][:-1] # Remove ":"

		# Filter the verbs
		verbs = self.dictionary.verbs(strVerb)

		# Verb not found
		if not verbs:
			self.debug("verbs not found.")
			
			# Check if the token is an exit
			exit = self.dictionary.exit(line.strip())
			if exit:
				self.debug(f'exit: "{exit.name}" found.')
				# Get the verb associated to GoTo action
				gotoVerb = self.dictionary.verbByAction("GoTo")
				self.__dObjStr = line.strip()
				self.__iObjStr = ""

				# Instantiate the GoTo action
				action = gotoVerb.action()
				action.verb = gotoVerb
				action.game = self.game

				# Execute
				self.debug(f'calling GoTo action on exit: "{self.__dObjStr}".')
				action.execute()
				
			else:
				self.debug("exit not found.")
				self.parse("?")
			return None

		self.debug(f"for {tokens[0]}, {len(verbs)} verb(s) found, checking syntax ...")

		action = None

		# Checks the verbs syntaxs
		for v in verbs:
			# Clear objects string
			self.__dObjStr = ""
			self.__iObjStr = ""
			self.__keyword = ""
			self.__parameters = ""

			action = self.checkSyntax(v, tokens)
			if action: break # Syntax not match

		if not action:
			self.debug(f'syntax check fails: "{tokens[0]}".')

			# Print syntax fail response on first verb (print response)
			response = verbs[0].getResponse("syntax-fail").strip()
			if response:
				Console.println(response)
			else:
				self.game.execute("?")


			return None

		# Clean articles from object strings
		self.__dObjStr = self.cleanArticles(self.__dObjStr)
		self.__iObjStr = self.cleanArticles(self.__iObjStr)

		self.debug(f'executing action: "{action.__class__}".')

		# Debug: show the parameters
		if self.__showParsingProcess:
			if self.__dObjStr:
				msg = f"Params 1={self.__dObjStr}"
				if self.__iObjStr:
					msg += f", 2={self.__iObjStr}."
				self.debug(msg)

		action.execute()


	def checkSyntax(self, verb: "dragonfly.Verb", tokens: typing.List[str]) -> "dragonfly.Action":
		# Instantiate the verb's action
		actionClass = verb.action
		action = actionClass()
		action.game = self.__game
		action.verb = verb

		syntax = verb.syntax

		# CASE 0: Multiparameter verb
		if syntax:
			if syntax[-1] == "...":
				self.debug("Multiparameter verb:")
				return self.checkMultiparameterVerb(action, tokens)

		# CASE 1: Verb without parameters
		if not syntax:
			if len(tokens) > 1: return None
			return action

		# CASE 2: Wait for direct object
		if len(syntax) == 1:
			if syntax[0] == "1":
				# Token is only the verb ...
				if len(tokens) == 1:
					# Requires a direct object
					return None
				
				# Exclude the verb
				self.__dObjStr = " ".join(tokens[1:])
				return action
			else:
				raise dragonfly.DragonflyException(f"Syntax error on verb: {verb.name}.")

		# CASE 3: Wait keyword and direct object
		if len(syntax) == 2:
			if syntax[0] == "1" or syntax[0] == "2":
				raise dragonfly.DragonflyException(f"Syntax error on verb: {verb.name}.")

			if syntax[1] == "1":
				if len(tokens) <= 1: return None

				# Check the keyword
				if self.checkKeyword(tokens[1], syntax[0]):
					self.__keyword = tokens[1]
					self.__dObjStr = " ".join(tokens[2:])
					return action
				else:
					# Keyword not found
					return None

			else:
				raise dragonfly.DragonflyException(f"Syntax error on verb: {verb.name}.")

		# CASE 4: Wait for 3 things
		if len(syntax) == 3:
			# CASE 4A: Object, keyword, object
			if syntax[1] != "1" and syntax[1] != "2":
				if (syntax[0] == "1" and syntax[2] == "2") or (syntax[0] == "2" and syntax[2] == "1"):
					# Read the first object until the keyword
					ti = 1
					while ti < len(tokens) and not self.checkKeyword(tokens[ti], syntax[1]):
						if syntax[0] == "1":
							self.__dObjStr += tokens[ti] + " "
						else:
							self.__iObjStr += tokens[ti] + " "
						ti += 1

					# Keyword not found
					if ti >= len(tokens): return None

					# Set the keyword
					self.__keyword = tokens[ti]

					# Now, search for the other object
					ti += 1
					while ti < len(tokens):
						if syntax[2] == "1":
							self.__dObjStr += tokens[ti] + " "
						else:
							self.__iObjStr += tokens[ti] + " "
						ti += 1

					# If the object is null
					if syntax[2] == "1":
						if not self.__dObjStr: return None
					else:
						if not self.__iObjStr: return None

					# Ok
					return action
				else:
					raise dragonfly.DragonflyException(f"Syntax error on verb: {verb.name}.")

			# CASE 4B: Keyword, object, object
			if syntax[0] != "1" and syntax[1] != "2":
				# Expected keyword first
				if not self.checkKeyword(tokens[1], syntax[0]):
					return None

				# Search for the first object until an article definited in dictionary
				ti = 2
				while ti < len(tokens) and not self.game.dictionary.article(tokens[ti]):
					if syntax[1] == "1":
						self.__dObjStr += tokens[ti] + " "
					else:
						self.__iObjStr += tokens[ti] + " "

					ti += 1

				# Article not found
				if (ti >= len(tokens)): return None

				# Search for the second object
				while ti < len(tokens):
					if syntax[2] == "1":
						self.__dObjStr += tokens[ti] + " "
					else:
						self.__iObjStr += tokens[ti] + " "

					ti += 1

				# OK
				self.__keyword = tokens[1]
				return action

			
	def checkKeyword(self, keyword: str, kwList: str) -> bool:
		return keyword in kwList.split("/")

	def cleanArticles(self, obj: str) -> str:
		result = []
		words = obj.split(" ")
		for w in words:
			if self.game.dictionary.article(w): continue
			result.append(w)

		return " ".join(result)

	def debug(self, msg: str) -> None:
		if self.__showParsingProcess:
			Console.println(f"Parser: {msg}", "family: 'Courier'")

	def checkMultiparameterVerb(self, action: "dragonfly.Action", tokens: typing.List[str]) -> "action.Action":
		"""Check multiparameter sentence like talk action

		Args:
			action (action.Action): Action instance
			tokens (typing.List[str]): Sentence tokens

		Returns:
			action.Action: Action instance if and only if tokens are valid.
		"""
		verb = action.verb
		syntax = verb.syntax

		joined = " ".join(tokens)
		pair = joined.split(":")

		# ":" separator not found
		if len(pair) != 2: return False

		# CASE 1: Verb with only multiparameters
		if len(syntax) == 1:
			if not verb.responds(pair[0]):
				# Maybe a direct object found
				return False
			self.__parameters = pair[1].strip()
			return action

		# CASE 2: Verb with keyword, direct object and multiparameters
		if len(syntax) == 3:
			leftTokens = pair[0].split(" ")

			# Expected direct object indicator
			if syntax[1] != "1":
				raise dragonfly.DragonflyException(f"Syntax error on verb: {verb.name}.")

			# Not enough elements
			if len(leftTokens) < 3: return None

			# Search for a keyword
			if not self.checkKeyword(leftTokens[1], syntax[0]): return None

			self.__keyword = leftTokens[1]

			# Get direct object
			for tk in range(2, len(leftTokens)):
				self.__dObjStr += leftTokens[tk]

			# Get multiparameters
			self.__parameters = pair[1].strip()

			return action
			

File: base/parser.js
import { Action } from "./action.js";
import { Output } from "./output.js";
import { Verb } from "./verb.js";
import { Book } from "./book.js";
import { Utils } from "./utils.js";

/**
 * Main parser module
 *
 * @export
 * @class Parser
 */
export class Parser {
	/**
	 * Creates an instance of Parser.
	 * @param {Book} book main book instance.
	 * @memberof Parser
	 */
	constructor(book) {
		this.book = book;
		this.showParsingProcess = false;

		this.directObjectString = "";
		this.indirectObjectString = "";
		this.parameters = "";
		this.directObject = null;
		this.indirectObject = null;
		this.keyword = "";
	}

	/**
	 * Main parse method.
	 *
	 * @param {string} line string line to parse.
	 * @memberof Parser
	 */
	async parse(line) {

		// Dragonfly version
		if (line.trim().toLowerCase() === "--version") {
			Utils.showBookInfo(this.book);
			return;
		}

		// Clean unexpected chars
		const regex = new RegExp(`[${this.book.getProperty("parser-clean")}]`, 'g');
		line = line.replace(regex, '');

		const tokens = line.trim().split(" ").filter(t => t.trim());

		// Empty line
		if (!tokens.length) {
			this.debug("token list is empty.");
			return;
		}

		let strVerb = tokens[0];

		if (strVerb.endsWith(":")) {
			strVerb = strVerb.slice(0, -1); // Remove ":"
		}

		// Filter the verbs
		const verbs = this.book.dictionary.getVerbs(strVerb);

		// Verb not found
		if (!verbs.length) {
			this.debug("verbs not found.");
			
			// Check if the token is an exit
			const exit = this.book.dictionary.getExit(line.trim());
			if (exit) {
				this.debug(`exit: "${exit.getName()}" found.`);
				const gotoVerb = this.book.dictionary.verbByAction("GoTo");
				this.directObjectString = line.trim();
				this.indirectObjectString = "";

				const action = new gotoVerb.action();
				action.verb = gotoVerb;
				action.book = this.book;

				this.debug(`calling GoTo action on exit: "${this.directObjectString}".`);
				await action.execute();
			} else {
				this.debug("exit not found.");
				this.parse("?");
			}
			return;
		}

		this.debug(`for ${tokens[0]}, ${verbs.length} verb(s) found, checking syntax ...`);

		let action = null;

		for (const v of verbs) {
			this.directObjectString = "";
			this.indirectObjectString = "";
			this.keyword = "";
			this.parameters = "";

			action = this.checkSyntax(v, tokens);
			if (action) break; // Syntax match found
		}

		if (!action) {
			this.debug(`syntax check fails: "${tokens[0]}".`);

			const response = verbs[0].getResponse("syntax-fail").trim();
			if (response === "") {
				Output.print(response);
			} else {
				this.book.execute("?");
			}
			return;
		}

		this.directObjectString = this.cleanArticles(this.directObjectString);
		this.indirectObjectString = this.cleanArticles(this.indirectObjectString);

		this.debug(`executing action: "${action.constructor.name}".`);

		if (this.showParsingProcess) {
			if (this.directObjectString) {
				let msg = `Params 1=${this.directObjectString}`;
				if (this.indirectObjectString) {
					msg += `, 2=${this.indirectObjectString}.`;
				}
				this.debug(msg);
			}
		}

		await action.execute();
	}

	/**
	 * Compare syntax verb and try detect direct and indirect objects. Then, returns
	 * the action associated to verb and syntax.
	 *
	 * @param {Verb} verb target verb.
	 * @param {Array<string>} tokens parsed token list.
	 * @return {Action} the action detected.
	 * @memberof Parser
	 */
	checkSyntax(verb, tokens) {
		const action = new verb.action();
		action.book = this.book;
		action.verb = verb;

		const syntax = verb.syntax;

		// Case 0: Multiparameter verb
		if (syntax.length !== 0 && syntax[syntax.length - 1] === "...") {
			this.debug("Multiparameter verb:");
			return this.checkMultiparameterVerb(action, tokens);
		}

		// Case 1: Verb without parameters
		if (syntax.length === 0) {
			if (tokens.length > 1) return null;
			return action;
		}

		// Case 2: Wait for direct object
		if (syntax.length === 1 && syntax[0] === "1") {
			if (tokens.length === 1) return null;
			this.directObjectString = tokens.slice(1).join(" ");
			return action;
		}

		// Case 3: Keyword and direct object
		if (syntax.length === 2 && syntax[1] === "1") {
			if (tokens.length <= 1) return null;
			let result = this.checkKeyword(1, tokens, syntax[0]);
			if (result) {
				this.keyword = result.keyword;
				this.directObjectString = tokens.slice(result.end).join(" ");
				return action;
			} else {
				return null;
			}
		}

		// Case 4: Object, keyword, object
		if (syntax.length === 3) {
			if (syntax[1] !== "1" && syntax[1] !== "2") {
				let ti = 1;
				let result = null;
				while (ti < tokens.length && (result = this.checkKeyword(ti, tokens, syntax[1])) === null) {
					if (syntax[0] === "1") {
						this.directObjectString += tokens[ti] + " ";
					} else {
						this.indirectObjectString += tokens[ti] + " ";
					}
					ti++;
				}

				if (ti >= tokens.length) return null;
				this.keyword = result.keyword;

				ti = result.end;

				while (ti < tokens.length) {
					if (syntax[2] === "1") {
						this.directObjectString += tokens[ti] + " ";
					} else {
						this.indirectObjectString += tokens[ti] + " ";
					}
					ti++;
				}

				if (!this.directObjectString && syntax[2] === "1") return null;
				if (!this.indirectObjectString && syntax[2] !== "1") return null;

				return action;
			}
		}

		throw new Error(`Syntax error on verb: ${verb.name}.`);
	}

	/**
	 * Check if the keyword is in keyword list.
	 *
	 * @param {number} start the start index.
	 * @param {Array<string>} the token list.
	 * @param {string} keywords the keyword list.
	 * @return {dictionary} null if not found. For otherwise, the keyword and the end index.
	 * @memberof Parser
	 */

	checkKeyword(start, tokens, keywords) {

		// Separate keywords
		for (const kw of keywords.toLowerCase().split("/")) {

			let index = start;
			let match = true;

			// Check word by word
			for (const w of kw.split(" ")) {
				if (tokens[index].toLowerCase() !== w) {
					match = false; break;
				}
				index ++;
			}

			if (match) {
				return {
					// Join the keyword
					keyword: tokens.slice(start, index).join(" "),
					end: index // End index
				};
			}
		}

		return null;
	}

	/**
	 * Removes the dictionary registered articles from given object.
	 *
	 * @param {string} obj the string name of the object.
	 * @return {string} the result without the articles.
	 * @memberof Parser
	 */
	cleanArticles(obj) {
		return obj.split(" ").filter(w => {
			const wordLower = w.toLowerCase();
			return !this.book.dictionary.getArticle(wordLower);
		}).join(" ").trim();
	}

	/**
	 * Check the multiparameter special verb and returns the associated action.
	 *
	 * @param {*} action
	 * @param {*} tokens
	 * @return {*} 
	 * @memberof Parser
	 */
	checkMultiparameterVerb(action, completeTokens) {
		const verb = action.verb;
		const syntax = verb.syntax;

		const pair = completeTokens.join(" ").split(":");

		let tokens = "";
		let parameters = "";

		if (pair.length === 1) {
			this.debug("Separator ':' not found. No parameters.");
			tokens = pair[0];

		} else if (pair.length > 2) {
			this.debug("Error: Multiple separators ':' found.");
			return null;
		} else {
			tokens = pair[0].trim();
			parameters = pair[1].trim();
		}

		if (syntax.length === 1) {
			if (!verb.responds(tokens)) {
				this.debug(`verb: '${verb.getName()}' is not '${tokens}.'`);
				return null;
			}
			this.parameters = parameters;
			this.debug(`Parameters: '${parameters}'.`);
			return action;
		}

		if (syntax.length === 3) {
			const leftTokens = tokens.split(" ");

			if (syntax[1] !== "1") throw new Error(`Syntax error on verb: ${verb.name}.`);

			if (leftTokens.length < 3) return null;

			//if (!this.checkKeyword(leftTokens[1], syntax[0])) return null;
			const result = this.checkKeyword(1, leftTokens, syntax[0]);
			if (!result) return null;

			this.keyword = result.keyword;
			this.directObjectString = leftTokens.slice(result.end).join(" ");
			this.parameters = parameters;

			this.debug(`Keyword: '${this.keyword}'.`);
			this.debug(`Parameters: '${parameters}'.`);

			return action;
		}

		return null;
	}

	/**
	 * Print a debug message if 'showParsingProcess' is activated.
	 * 
	 * @param {string} msg debug message.
	 */
	debug(msg) {
		if (this.showParsingProcess) {
			Output.print(`Parser: ${msg}`, {
				fontFamily: "'Courier New', Courier, monospace",
				color: "#888"
			});
		}
	}
}
